prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt

```diff
@@ -40,6 +44,40 @@ class HL7Reader(private val actionLogger: ActionLogger) : Logging {
         return messages
     }
 
+    /**
+     * Takes an [exception] thrown by the HL7 HAPI library, gets the root cause and logs the error into [actionLogger].
+     * Sample error messages returned by the HAPI library are:
+     *  Error Code = DATA_TYPE_ERROR-102: 'test' in record 3 is invalid for version 2.5.1
+     *  Error Code = REQUIRED_FIELD_MISSING-101: Can't find version ID - MSH.12 is null
+     * This functions only logs messages that contain meaningful data.
+     *
+     */
+    private fun logHL7ParseFailure(exception: Hl7InputStreamMessageStringIterator.ParseFailureError) {
+        logger.error("Failed to parse message", exception)
+        // Get the exception root cause and log it accordingly
+        when (val rootCause = ExceptionUtils.getRootCause(exception)) {
+            is ValidationException -> {
+                actionLogger.error(InvalidReportMessage("Validation Failed ${rootCause.message}"))
+            }
+
+            is HL7Exception -> {
+                when (rootCause.errorCode) {
+                    ErrorCode.REQUIRED_FIELD_MISSING.code -> actionLogger.error(
+                        InvalidReportMessage("Required field missing: ${rootCause.message}")
+                    )
+                    ErrorCode.DATA_TYPE_ERROR.code -> actionLogger.error(
+                        InvalidReportMessage("Data type error: ${rootCause.message}")
+                    )
+                    else -> actionLogger.error(InvalidReportMessage("Failed to parse message"))
+                }
+            }
+
+            else -> {
+                actionLogger.error(InvalidReportMessage("Failed to parse message"))
+            }
```

Comment:
(Optional) An option to make this function more concise would be to initialize a `var errMessage = "InvalidReportMessage("Failed to parse message")` or `var errMessageStr = "Failed to parse message"` at the top, then update that var if we have a more specific message, and log that `errMessage` either way at the end. It would allow us to get rid of the duplicate else cases at 76 and 71, but would involve a tiny bit of extra operations by having to init and overwrite the var in some cases.

Reply:
I tried making that change, but `when` requires an else statement... I did turn the error message into a variable to make the function more readable.